### PR TITLE
feat: Tree Select widget

### LINKED ISSUE #1759: [Feature] Nested Tree Selector Widget
# RACI matrix
| Label | Value | 
| ------------- | ------------- |
| Responsible |  @momcilo-appsmith |
| Accountable | @somangshu  |
| Consulted |  |
| Informed |  @Nikhil-Nandagopal |

# Objective

We need a tree selector widget that allows users to select single and multiple values from a list of options.

> TreeSelect is similar to Select, but the values are provided in a tree like structure. Any data whose entries are defined in a hierarchical manner is fit to use this control. Examples of such case may include a corporate hierarchy, a directory structure, and so on.

# Success Metrics

| Goal  | Metric |
| ------------- | ------------- |
| New widget  | Increase use case coverage |
| Nested objects  | ability to handle nested objects inside Appsmith |

# Requirements

| Requirement   | User Story | Importance | Notes |
| ------------- | ------------- | ------------- | ------------- |
| Type | Weather to allow single or multiple selection | High | - |
| During single select show radio as a selection | When the dev chooses single select mode the user will be able to see radio buttons with the options since only one option can be selected | Medium  | - |
| During multiple select show checkbox as a selection | When the dev chooses multi select mode the user will be able to see checkbox buttons with the options since many option can be selected | Medium  | - |
| Options / Tree Data  | contain nested collection of Tree select children `{"label": "", "value": "", "children": [{}]}` | High  | - |
| isVisible |  Dev can set this attributes to show/hide with toggle or even a JS expression (JS convertible)  | High  | - |
| isDisabled | Dev can set this attributes to disable the widget with toggle or even a JS expression (JS convertible) | High | - |
| Action - `onChange` | API to trigger an action on change or se[de]lection of a option | High  | - |
| allowClear | Whether allow clear : boolean | High  | - |
| maxTagCount | Internally fixed to responsive, Tags will cover the space give and then show `+2 more` | High  | - |
| Placeholder | Placeholder of the tree input | Medium | - |
| Internal property: value | Get all selected values (Array is multiple); string : string[] | High | - |
| treeDefaultExpandAll | Whether to expand all treeNodes by default | Low | - | 
| Default Options | The dev can use this prop to set the default values which will be selected when the widget is loaded. The type for this field is `<array>` and we will do a deep search in the main object | High | - | 

# Questions:

| Question   | Answer |
| ------------- | ------------- |
| Imagine there is a parent item in a multiselect mode that has 2 child items. What if the user selects the parent item and one child item? Do we show both of them in the input or just the parent one |  |

# Developer Handoff Document in Figma:
[V1](https://www.figma.com/file/hsHKmSVcoQVKOoTXOIs7rQ/?node-id=136%3A1527) (changes required here)

[Final Design](https://www.figma.com/file/1HLLnIw4Fudx8EpgouaEDU/TreeSelect-Widget-Tue-20-July?node-id=2%3A1)

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

New Tree Select Widget

Fixes #1759 

## Type of change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feat/Tree-Select-Widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.79 **(-0.1)** | 36.47 **(-0.25)** | 34.07 **(-0.13)** | 55.33 **(-0.09)**
 :red_circle: | app/client/src/icons/WidgetIcons.tsx | 61.43 **(-0.69)** | 100 **(0)** | 20.59 **(-1.29)** | 61.43 **(-0.69)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.47 **(-0.24)** | 38.26 **(-0.87)** | 36.21 **(0)** | 55.08 **(-0.27)**
 :red_circle: | app/client/src/utils/validation/common.ts | 22.41 **(0)** | 15.72 **(-0.1)** | 33.33 **(0)** | 21.05 **(0)**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectTreeWidget/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectTreeWidget/component/index.styled.tsx** | **54.84** | **26.92** | **0** | **55.56**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx** | **20.45** | **0** | **0** | **23.08**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx** | **45** | **0** | **42.86** | **46.43**
 :sparkles: :new: | **app/client/src/widgets/SingleSelectTreeWidget/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/SingleSelectTreeWidget/component/index.styled.tsx** | **58.62** | **28** | **0** | **57.69**
 :sparkles: :new: | **app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx** | **20.93** | **0** | **0** | **23.68**
 :sparkles: :new: | **app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx** | **53.85** | **0** | **50** | **56.25**
 :red_circle: | app/client/src/workers/validations.ts | 73.29 **(-5.86)** | 65.71 **(-3.25)** | 63.33 **(-9.75)** | 75.36 **(-5.41)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@Tooluloope the names of the widget can be as follows
- Single select tree
- Multi select tree

File: app/client/src/components/designSystems/appsmith/TreeSelectComponent/SingleSelectComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+      )}
+      <TreeSelect
+        allowClear={allowClear}
+        animation="slide-up"
+        choiceTransitionName="rc-tree-select-selection__choice-zoom"
+        className="rc-tree-select"
+        disabled={disabled}
+        dropdownClassName="tree-select-dropdown single-tree-select-dropdown"
+        dropdownStyle={dropdownStyle}
+        getPopupContainer={() => getDropdownPosition(_menu.current)}
```
Comment (jsartisan) [inline app/client/src/components/designSystems/appsmith/TreeSelectComponent/SingleSelectComponent.tsx:136]: the arrow function here defeats the purpose of useCallback.

Here it should be like:
```
getPopupContainer={getDropdownPosition}
```

and the callback should be defined as

```
 const getDropdownPosition = useCallback(() => {
    const node = _menu.current;
    if (Boolean(node?.closest(`.${MODAL_PORTAL_CLASSNAME}`))) {
      return document.querySelector(
        `.${MODAL_PORTAL_CLASSNAME}`,
      ) as HTMLElement;
    }
    return document.querySelector(`.${CANVAS_CLASSNAME}`) as HTMLElement;
  }, []);
  ```

File: app/client/src/components/designSystems/appsmith/TreeSelectComponent/SingleSelectComponent.tsx [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        clipRule="evenodd"
+        d="M2.5 9L7.5 5L2.5 1L2.5 9Z"
+        fill="#090707"
+        ill-rule="evenodd"
+      />
+    </svg>
+  </i>
+);
+
+const switcherIcon = (obj: TreeNodeProps) => {
```
Comment (jsartisan) [inline app/client/src/components/designSystems/appsmith/TreeSelectComponent/SingleSelectComponent.tsx:67]: Can we avoid using variable names like obj and use descriptive names. May be we can use the name here: treeNode

File: app/client/src/components/designSystems/appsmith/TreeSelectComponent/SingleSelectComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        dropdownStyle={dropdownStyle}
+        getPopupContainer={() => getDropdownPosition(_menu.current)}
+        inputIcon={inputIcon}
+        loading={loading}
+        maxTagCount={"responsive"}
+        maxTagPlaceholder={(e) => `+${e.length} more`}
+        notFoundContent="No item Found"
+        onChange={onChange}
+        onClear={() => {
+          onChange([], []);
```
Comment (jsartisan) [inline app/client/src/components/designSystems/appsmith/TreeSelectComponent/SingleSelectComponent.tsx:144]: Similarly the use of the arrow function in this defeats the purpose of usecallback because it is creating function on every render.

File: app/client/src/components/designSystems/appsmith/TreeSelectComponent/index.styled.tsx [hunk 4]
```diff
@@ -0,0 +1,894 @@
+import React from "react";
```
Comment (jsartisan) [inline app/client/src/components/designSystems/appsmith/TreeSelectComponent/index.styled.tsx:1]: This is not a bug but just wanted to know if index.styled.tsx is a naming convention of styled-components files?

File: app/client/src/components/designSystems/appsmith/TreeSelectComponent/index.styled.tsx [hunk 5]
```diff
@@ -0,0 +1,894 @@
...
...uUC9aYIs+4Yi2yE3Ga5qggIBWrVZPz8jIkOGB9/EcLzruJmAXtcDq6NDG8VVHS3o6VuKAQjPAH+cHJiFZ72kJqbAy1F3kmEYeTyDeb1ZqamoyrvHwwPt4DtuwD/ZFDGK7p+AuYjdQHb3ovQWZoBddKGkm8UGJOwR4dV4m/HFDIV/Pb7HI6w0KDw//Ii4uTo3Bh9VqZTTEBg4cGNvQwF17jvdJgPKujZhWq1WgFzQYDPaWlha88Ol0NwG7gN1IdXQx4cmFAPGmiawIXpydCW9v8iVhZWWlMyIiIpas92KSkpLoD1objUbiee3AE1Cn0ymys7OTSD/6W861tbWwffv2JsR2e8BuAzMhWKvZfzsVVRGP+JcHM+HZzwq9yrLt3r27mEyzz5rN5oUTJkzIwd8cQRIS7+ZZ7yEho6Ki6I+Jnz59mj18+PDR0tLS1fv37y/uJmC3gYXEJiYz47ddp1ZAShgg+cBhbvmHl3c0mezEm/2LTMMlly5dWjJjxox7evXqpcRUjM39K5xIPAxAvvvuOyfpu+PQoUPLCGGPkWnZ3k3AboM0HSFhtPelm612BqpbuURxZqIC1uwrhNbK0i8vvDrzKXjSK5JlCZFshIgHCgoKLH379h2QlpY2kKwFaXKaj44xSX3x4sVS0ud10vf49YyGuwl4E5u16er6d3bCfKm2H93WDyI0cvjnEQ/5Hsn5qMCnrgv+zFdCQgKMHz9ek5iYqMbIlwQbwO8Z81W3sC03N1dz5MgRqK+vx/VjNwF/6Hb6uTtRTvAazrTC84RoZ7J7quDNXYHJR4IPGDt2LAYdaqVSOblPnz49MdDA7bmioiLqAgcNGqTEilvYRqLfyWPGjMlXq9X2Y8eOdRPwh25uUpVKecY3d8H8QORDmzZtGqZesKxbSmRkZC7xcloMQI4ePVqTn5+/FfsQbzczJyenJ7bFxsbmtra2YiGkMsR2E7DbAnlG1P2Z/JEPrampiV/nqck6T028Wsu5c+f2HDhw4BPiBakekKz9tpSXlz+SlZU1lUTIahKc8DnD6/Jauy9M/wFbXFwcfxen4IHEyw2qrq4+3djYWNy7N/djj1euXAHi+fonJycPv3r1ahEJTlBhQyNgMiV3E7DbOvDh+9buwRmRrv2EQYi4zRNCXwfudBOw226o/Z8AAwBphnYirXZBiwAAAABJRU5ErkJggg==");
```
Comment (jsartisan) [inline app/client/src/components/designSystems/appsmith/TreeSelectComponent/index.styled.tsx:486]: is this okay to use datauri here? dataruri images/files actually make the filesize big but save HTTP calls. It's to up to you to change this.

File: app/client/src/components/designSystems/appsmith/TreeSelectComponent/index.styled.tsx (hunk 4 above)
Comment (Tooluloope) [reply app/client/src/components/designSystems/appsmith/TreeSelectComponent/index.styled.tsx:1]: @jsartisan Didn't see any convention, that's why I used that naming style

**Review by jsartisan — APPROVED** [review]

File: app/client/src/constants/WidgetValidation.ts [hunk 6]
```diff
@@ -10,6 +10,7 @@ export enum ValidationTypes {
   OBJECT = "OBJECT",
   ARRAY = "ARRAY",
   OBJECT_ARRAY = "OBJECT_ARRAY",
+  NESTED_OBJECT_ARRAY = "NESTED_OBJECT_ARRAY",
```
Comment (Tooluloope) [inline app/client/src/constants/WidgetValidation.ts:13]: @jsartisan , I added a new validation for the Nested Option, the previous one, seems to be breaking in some edge cases

**Review by jsartisan — DISMISSED** [review]

**Comment by somangshu** [comment]
@Tooluloope please resolve the conflicts

**Review by somangshu — APPROVED** [review]
Approving for tests to run

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Tested this PR and please find the status below,

1.Both MultiSelectTree and SingleSelectTree - Accepts duplicate value and label

Screenshot 2021-08-24 at 3 51 49 PM

2. The icons in the entity explorer for ''MultiselectTree' and 'SingleSelectTree' are same

Screenshot 2021-08-25 at 1 44 21 PM

3. Both MultiSelectTree and SingleSelectTree - When the Expand all by default option is disabled - The dropdown options are not displayed by default only after refreshing the page 

https://loom.com/share/45036a02655d4872a1b9782c6d0eafe6

4.SingleSelectTree - Radio buttons are very light and can be little more darker 

Screenshot 2021-08-25 at 1 52 49 PM

5.MultiSelectTree - Label text breaks when the label text size is chosen as 'Heading 1'

Screenshot 2021-08-25 at 1 56 26 PM

6.Both MultiSelectTree and SingleSelectTree - When the options value are removed - 'undefined' is displayed for children value and which should be 'Value' instead of undefined

Screenshot 2021-08-25 at 1 57 20 PM

7.Both MultiSelectTree and SingleSelectTree - If the widgets size is reduced so that the Label text comes next to widget and when placed inside the form widget - The label text and the values overlaps inside form widget 

https://www.loom.com/share/a41d9b4a8ad748e881b5fff73ddf2c98

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Tested the above fixes and please find the status below,

**Issues Fixed:**
2,4,6

**Default Behaviour:**
Agreed upon issue 5 and 7 

**Re-opening issues:**

1. OnEntering Duplicate value in the options field - 'This Value does not evaluate to type "array" error occurs in the widget - But in the debugger appropriate error 'Value cannot be duplicate' is displayed - Error message should be unique both at widget and debugger 

https://loom.com/share/986b452830ad4424b28591db793f389c

3. Enabling and Disabling - Expand all by default - Still does not work and works only after refreshing page 
https://loom.com/share/61d7e574c07541ec8e6284236879c274

**New issues:**

8. Single Select Tree - Default value is not selected in the widget 

https://loom.com/share/9cd9195cf72440b5b3c705bba38512e5

9.Multi Select Tree - By selecting few more options and unchecking some the widget shows improper UI - Widget outside container is not appropriate 

https://loom.com/share/743707d5ecd742abb6bc22059755dbfe

**Comment by Tooluloope** [comment]
@YogeshJayaseelan 
1. is been fixed here https://github.com/appsmithorg/appsmith/issues/5840
8. Fixed
9. Fixed

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Verified the fixes for issue 3, 8 and 9 and they are working as expected. We can proceed further

**Review by jsartisan — DISMISSED** [review]

**Review by jsartisan — APPROVED** [review]
